### PR TITLE
"MegaGDN" kernel enabling 15% faster prefill for Qwen3.5/3.6 models

### PR DESCRIPTION
**Pull request by learning-chip**
### What this PR does / why we need it?

**3x faster** kernels than existing `chunk_gated_delta_rule` inside [vllm_ascend/ops/triton/fla](https://github.com/vllm-project/vllm-ascend/tree/v0.19.1rc1/vllm_ascend/ops/triton/fla). Measured **15% prefill speed-up (15% less TTFT)** for various Qwen3.5/3.6 models. No accuracy impact as shown by lm-eval. See our [full report](https://github.com/huawei-csl/megagdn-pto/tree/main/blog).

To verify the benefit of this new feature, it's easier to first use the [our "patch" style repo `megagdn-pto`](https://github.com/huawei-csl/megagdn-pto), to avoid rebuilding vllm-ascend from scratch. This PR brings such external patch into vllm-ascend source, for better long-term maintainability.

### Does this PR introduce _any_ user-facing change?

No, just a faster internal kernel.

### How was this patch tested?

The original standalone [`megagdn-pto` repo](https://github.com/huawei-csl/megagdn-pto) has test scripts for:
- Single-kernel unit tests and benchmarks
- Prefill TTFT benchmark
- lm-eval score evaluation

I can put some of them into this PR. Let me know which are needed.

### Notes for vllm-ascend maintainers

This is the first attempt to merge PTO-ISA kernels into vllm-ascend, so likely need a few rounds of reviews and edits. 

The current way this PR is structured:
- PTO C++ kernels are in `csrc/pto_chunk_gdn/`, the same level of directory as other AscendC kernel sources
- The torch interfaces are under `vllm_ascend/ops/pto_chunk_gdn/`
- I put a feature guide `docs/source/user_guide/feature_guide/enable_megagdn_kernel.md`
- To provide PTO-ISA dependency, I added https://gitcode.com/cann/pto-isa to `csrc/third_party/pto-isa` as git submodule, the same way as the `csrc/third_party/catlass` submodule.
- CANN installation also provides PTO-ISA headers, similar to how it provides AscendC headers. It's located in `$ASCEND_HOME_PATH/include/pto`. But the headers in 8.5.0 is not new enough, thus I pulled a more recent commit. With future newer CANN releases, this `pto-isa` git submodule might not be needed.

Let me know your preferred: (1) code organization (2) compilation procedure (3) if more docs and tests needed. Then we can make further changes to make this PR more "merge-able".

I assume the PTO torch interface needs to be the same as current AscendC build structure? i.e. Using CMake + pybind (torch C++ api). An alternative without needing to rebuild csrc directory, is to use "JIT compile" style, calling `bisheng` + `ctypes` on the fly, so it's more like triton/tilelang's way of invoking kernels.

### Known uncovered cases by this PR

- Kernels are fully tested on Atlas A2, and should also be compatible with A3 (same aicore isa), but I didn't verify directly (no A3 machine)
- A5 & Ascend 950 is not tested *yet*. Although PTO-ISA abstracts away A3 vs A5 differences, we still need slight tweaks to the kernel sources and launch interface (mostly for Cube-Vector communication). This is planned as the next PR (current one is already big enough, don't want to bloat the code)
- Multi-device inference for larger 122B/397B models are not tested yet. Will need to fuse communication to gain best speed-up, otherwise the eager communication calls will kill any single-kernel speed-up. This is planned as future PRs.

## Extended summary 

(below are AI summarized but quite correct; above are purely human written)

<details>

PR: Enable PTO GDN Megakernel for Qwen3.5 / Qwen3.6 prefill

**Target**: `vllm-project/vllm-ascend` main branch

---

## Summary

Integrate fused PTO megakernel for the **chunk GatedDeltaNet (GDN)** used in Qwen3.5 and
Qwen3.6 models, replacing the default Triton implementation during prefill.

Key outcomes:
- **7–25% TTFT reduction** on Ascend 910B, depending on model size and sequence length.
- **Zero accuracy regression**: WikiText PPL and MMLU scores match Triton to floating-point equivalence.
- **Decode unaffected**: the PTO path only activates for prefill (zero `initial_state`); decode transparently falls back to Triton.
- **Single opt-in env var**: `VLLM_ASCEND_PTO_CHUNK_GDN=1`.

---

## Background

Qwen3.5 (0.8B – 9B) and Qwen3.6 (27B, 35B-A3B-MoE) use a **GatedDeltaNet**
recurrent layer that dominates prefill cost at long context lengths.  The
existing Triton-based implementation runs with `chunk_size=64` and BF16.
The new PTO megakernel runs with `chunk_size=128` in FP16 on the Ascend
AI-Core vector unit, fusing all six pipeline stages into a **single NPU dispatch**.

### Pipeline stages (all fused into one kernel launch)

```
Q, K, V, g, β, cu_seqlens
│
├── cumsum       prefix cumulative sum of gate logits g        [Vec]
├── kkt          intra-chunk K^T K attention matrix A          [Cube+Vec]
├── solve_tril   lower-triangular solve A_inv = tril(A)^{-1}  [CubeCore]
├── wy_fast      Woodbury updates W, U from K, V, β, A_inv    [Vec+Cube]
├── chunk_h      inter-chunk recurrent state S and V'          [Cube+Vec]
└── chunk_o      output O = Q (K^T V' + S)                    [Cube+Vec]
```

GQA is fully supported: Q/K use `Hg` heads while V/gates use `H ≥ Hg` value
heads.

---

### New: `csrc/pto_chunk_gdn/`

| File | Description |
|---|---|
| `mega_kernel.cpp` | Fused megakernel entry point (all six stages) |
| `chunk_cumsum.cpp` | Cumulative sum stage (included by mega_kernel.cpp) |
| `scaled_dot_kkt.cpp` | K^T K attention matrix (included by mega_kernel.cpp) |
| `tri_inverse_impl.cpp` | Triangular inverse implementation (included by mega_kernel.cpp) |
| `wy_fast.cpp` | Woodbury W/U updates (included by mega_kernel.cpp) |
| `chunk_h.cpp` | Inter-chunk recurrent state (included by mega_kernel.cpp) |
| `chunk_o.cpp` | Output projection (included by mega_kernel.cpp) |
| `include/kernel_utils.h` | Shared AI-Core utility templates |
| `CMakeLists.txt` | Optional `BUILD_PTO_CHUNK_GDN` target for pre-compilation |

### New: `csrc/third_party/pto-isa` (git submodule)

PTO ISA header library required by the megakernel.  Added at the same level
as the existing `catlass` submodule:
```
csrc/third_party/pto-isa   ← new  (gitcode.com/cann/pto-isa.git)
csrc/third_party/catlass    ← existing
```

### New: `vllm_ascend/ops/pto_chunk_gdn/`

| Module | Description |
|---|---|
| `compile.py` | Bisheng JIT compilation; resolves `PTO_LIB_PATH`; caches `.so` |
| `mega_kernel.py` | ctypes launcher for the fused megakernel |
| `chunk_gated_delta_wrapper.py` | Drop-in `chunk_gated_delta_rule` replacement; handles GQA and Triton fallback for decode |
| `worker_hook.py` | `apply_pto_gdn_patch()` — patches `chunk_gated_delta_rule` at worker startup |

### Modified: `vllm_ascend/envs.py`

One new environment variable:

```python
VLLM_ASCEND_PTO_CHUNK_GDN   # bool, default False; enable PTO megakernel
```

### Modified: `vllm_ascend/patch/worker/__init__.py`

Native activation hook — runs after Triton patches, before model loading:

```python
if not is_310p():
    if envs.VLLM_ASCEND_PTO_CHUNK_GDN:
        from vllm_ascend.ops.pto_chunk_gdn.worker_hook import apply_pto_gdn_patch
        apply_pto_gdn_patch()
```

### Modified: `vllm_ascend/ops/gdn.py`

Changed static import to module-level attribute access so the worker hook
patch applies correctly:

```python
# Before
from vllm_ascend.ops.triton.fla.chunk import chunk_gated_delta_rule

# After
import vllm_ascend.ops.triton.fla.chunk as _fla_chunk_mod
# ...
(out, state) = _fla_chunk_mod.chunk_gated_delta_rule(...)
```

### New: `docs/source/user_guide/feature_guide/enable_megagdn_kernel.md`

Complete user guide covering quick-start, performance benchmarks, lm-eval
accuracy verification (with commands), CMake pre-compilation, and
troubleshooting.

---

## Performance Results

Measured on Ascend 910B4 with Qwen3.5-0.8B (BF16, eager mode, single device,
**rebuilt from source using this PR**):

| Backend | 512 tok TTFT | 1024 tok TTFT | 4096 tok TTFT |
|---|---|---|---|
| Triton (default)  | 125.8 ms | 125.3 ms | 133.8 ms |
| PTO megakernel    | 110.6 ms | 111.7 ms | 119.9 ms |
| **Speedup**       | **1.14×** | **1.12×** | **1.12×** |

For larger models (Qwen3.6-35B-A3B-MoE) where GDN layers dominate, speedups
reach **~25% at 4k+ tokens**.

---

## Accuracy Results (Qwen3.5-0.8B)

256-document wikitext subset + 6-subject MMLU
(**rebuilt from source using this PR**):

| Backend | WikiText PPL ↓ | MMLU acc ↑ |
|---|---|---|
| Triton (default) | 20.93 | 48.9% |
| PTO megakernel   | 20.93 | 48.9% |

Zero accuracy difference.

---

## Testing

```bash
# E2E inference: PTO and Triton produce identical outputs
ASCEND_RT_VISIBLE_DEVICES=0 \
VLLM_ASCEND_PTO_CHUNK_GDN=1 \
python -m vllm.entrypoints.openai.api_server \
    --model /path/to/Qwen3.5-0.8B --max-model-len 1024

# lm-eval accuracy
ASCEND_RT_VISIBLE_DEVICES=0 \
VLLM_ASCEND_PTO_CHUNK_GDN=1 \
python -m lm_eval --model vllm \
    --model_args "pretrained=/path/to/Qwen3.5-0.8B" \
    --tasks wikitext --limit 256
```

---

## Limitations

- Single-device only (TP=1, EP=1).  Multi-device support will be addressed in follow-up PRs.
- Chunk size fixed at C=128 (optimal tiling for Ascend 910B AI-Core).
- 910C/A3 uses the same code path but is not tested in the current CI environment.

---

## Notes for Reviewers

- Fully in-tree: no external dependencies beyond the `csrc/third_party/pto-isa` git submodule.
- `VLLM_ASCEND_PTO_CHUNK_GDN` is registered in `vllm_ascend/envs.py`; the
  "Unknown vLLM environment variable" warning from vLLM core is harmless.
- The JIT cache directory (`kernels/compiled_lib/*.so`) is listed in `.gitignore`.

---

### Prerequisites

- Ascend 910B device with CANN ≥ 8.0 and `bisheng` compiler in `PATH`
- Ubuntu 22.04 (recommended) with Python 3.11
- Activate the toolkit: `source /usr/local/Ascend/ascend-toolkit/set_env.sh`

### Build steps

```bash
# 1 — Clone with submodules
git clone https://github.com/vllm-project/vllm-ascend.git
cd vllm-ascend
git submodule update --init --recursive   # pulls pto-isa and catlass

# 2 — Install vLLM (empty device build — no GPU/CUDA needed)
git clone --depth 1 -b v0.19.1 https://github.com/vllm-project/vllm.git /tmp/vllm
VLLM_TARGET_DEVICE=empty pip install -e /tmp/vllm/ \
    --extra-index-url https://download.pytorch.org/whl/cpu/
pip uninstall -y triton

# 3 — Install vllm-ascend
pip install -e . \
    --extra-index-url https://mirrors.huaweicloud.com/ascend/repos/pypi \
    --extra-index-url https://download.pytorch.org/whl/cpu/
pip uninstall -y triton triton-ascend
pip install triton-ascend==3.2.0

# 4 — (Optional) Pre-compile PTO megakernel for common Qwen configs
cmake -S csrc -B csrc/build -DBUILD_PTO_CHUNK_GDN=ON
cmake --build csrc/build --target pto_chunk_gdn_kernels -j$(nproc)
```

### Run inference with PTO

```bash
ASCEND_RT_VISIBLE_DEVICES=0 \
VLLM_ASCEND_PTO_CHUNK_GDN=1 \
python -m vllm.entrypoints.openai.api_server \
    --model /path/to/Qwen3.5-7B \
    --max-model-len 8192
```

### Run accuracy check (lm-eval)

```bash
pip install lm-eval sacrebleu more-itertools datasets

# Triton baseline
ASCEND_RT_VISIBLE_DEVICES=0 python -m lm_eval \
    --model vllm \
    --model_args "pretrained=/path/to/Qwen3.5-0.8B,gpu_memory_utilization=0.85,enforce_eager=True" \
    --tasks "wikitext" --limit 256 --output_path results/triton.json

# PTO megakernel
ASCEND_RT_VISIBLE_DEVICES=0 VLLM_ASCEND_PTO_CHUNK_GDN=1 python -m lm_eval \
    --model vllm \
    --model_args "pretrained=/path/to/Qwen3.5-0.8B,gpu_memory_utilization=0.85,enforce_eager=True" \
    --tasks "wikitext" --limit 256 --output_path results/pto.json
```

---

## Minimum Dockerfile

A minimum `Dockerfile.pto` is included at the repository root for building a
self-contained image with CANN 8.5.1, vLLM v0.19.1, and vllm-ascend with PTO
pre-compiled for common Qwen configurations.

```dockerfile
FROM quay.io/ascend/cann:8.5.1-910b-ubuntu22.04-py3.11

ARG PIP_INDEX_URL="https://pypi.org/simple"
ARG VLLM_REPO="https://github.com/vllm-project/vllm.git"
ARG VLLM_TAG="v0.19.1"
ARG SOC_VERSION="ascend910b4"

WORKDIR /workspace

# System packages
RUN apt-get update -y && \
    apt-get install -y --no-install-recommends \
        git cmake ninja-build gcc g++ wget numactl libnuma-dev && \
    rm -rf /var/lib/apt/lists/*

RUN pip config set global.index-url ${PIP_INDEX_URL}

# vLLM (empty-device build)
RUN git clone --depth 1 -b ${VLLM_TAG} ${VLLM_REPO} /workspace/vllm && \
    VLLM_TARGET_DEVICE="empty" pip install -v -e /workspace/vllm/ \
        --extra-index-url https://download.pytorch.org/whl/cpu/ && \
    pip uninstall -y triton && pip cache purge

# vllm-ascend with pto-isa submodule
COPY . /workspace/vllm-ascend/
RUN cd /workspace/vllm-ascend && \
    git submodule update --init --recursive csrc/third_party/pto-isa

ENV SOC_VERSION=${SOC_VERSION} TASK_QUEUE_ENABLE=1 OMP_NUM_THREADS=1

RUN source /usr/local/Ascend/ascend-toolkit/set_env.sh && \
    source /usr/local/Ascend/nnal/atb/set_env.sh 2>/dev/null || true && \
    export PIP_EXTRA_INDEX_URL=https://mirrors.huaweicloud.com/ascend/repos/pypi && \
    pip install -v -e /workspace/vllm-ascend/ \
        --extra-index-url https://download.pytorch.org/whl/cpu/ && \
    pip uninstall -y triton triton-ascend 2>/dev/null || true && \
    pip install -v triton-ascend==3.2.0 && pip cache purge

# Pre-compile PTO megakernel (optional — remove to rely on JIT)
RUN source /usr/local/Ascend/ascend-toolkit/set_env.sh && \
    cmake -S /workspace/vllm-ascend/csrc \
          -B /workspace/vllm-ascend/csrc/build \
          -DBUILD_PTO_CHUNK_GDN=ON -DCMAKE_BUILD_TYPE=Release && \
    cmake --build /workspace/vllm-ascend/csrc/build \
          --target pto_chunk_gdn_kernels -j$(nproc) && \
    rm -rf /workspace/vllm-ascend/csrc/build

ENV VLLM_WORKER_MULTIPROC_METHOD=spawn \
    VLLM_ASCEND_PTO_CHUNK_GDN=1

RUN echo "source /usr/local/Ascend/ascend-toolkit/set_env.sh" >> ~/.bashrc

CMD ["/bin/bash"]
```

Build and run:

```bash
# Build image (from vllm-ascend repo root)
docker build --build-arg SOC_VERSION=ascend910b4 \
    -t vllm-ascend-pto:latest -f Dockerfile.pto .

# Run inference
docker run --rm -it \
    --device /dev/davinci0 --device /dev/davinci_manager \
    --device /dev/devmm_svm --device /dev/hisi_hdc \
    -v /usr/local/dcmi:/usr/local/dcmi \
    -v /usr/local/bin/npu-smi:/usr/local/bin/npu-smi \
    -v /usr/local/Ascend/driver/lib64:/usr/local/Ascend/driver/lib64 \
    -v /path/to/model:/model \
    vllm-ascend-pto:latest \
    python -m vllm.entrypoints.openai.api_server \
        --model /model --max-model-len 8192
```

- vLLM version: v0.19.1
- vLLM main: https://github.com/vllm-project/vllm/commit/d886c26d4d4fef7d079696beb4ece1cfb4b008a8

### CODE REVIEW

**Comment by mazhixin000** [comment]
Can this operator be used on A5?

**Comment by learning-chip** [comment]
> Can this operator be used on A5?

See https://github.com/huawei-csl/megagdn-pto/pull/14 it runs on A5 now, but performance not optimized yet (in our TODO)

**Comment by Sunshine-llh** [comment]
Hi, we are also seeing a very similar issue on our side with `Qwen3.5-35B-W8A8` on Ascend 910B:

  Issue reference: https://github.com/vllm-project/vllm-ascend/issues/11711

  ### What we observed
  - `Qwen3.5-35B-W8A8` is significantly slower than `Qwen3-30B-W8A8` when prefix caching is enabled.
  - Without prefix caching, `Qwen3.5-35B-W8A8` can even be faster than `Qwen3-30B-W8A8`.
  - The slowdown mainly appears in long-context, short-output workloads.
  - We also tried the GDN-related operator fusion / GDN prefill optimization path, but the overall behavior still looks
  abnormal for `Qwen3.5-35B`.

  ### Our main question
  Do you already know the root cause of the poor `Qwen3.5-35B` performance?

  In particular, is the main reason related to one or more of the following?
  1. experimental Mamba cache `align` mode when prefix caching is enabled
  2. GDN / DeltaNet execution path overhead
  3. GDN operator fusion not taking effect as expected
  4. extra state restore / state alignment cost for Qwen3.5 hybrid layers
  5. a model-specific implementation limitation for `Qwen3.5-35B` on Ascend

  ### Additional question about GDN fusion
  We also want to ask specifically about the GDN optimization direction:

  - Has the GDN operator fusion path already been fully validated for `Qwen3.5-35B`?
  - Does GDN fusion actually improve long-context prefix-cache scenarios for this model?
  - Are there known limitations or known regressions in the GDN prefill path for `Qwen3.5-35B`?
  - If you have already resolved this issue, could you share which patch / config / branch fixed it?

  Any update or workaround would be very helpful. Thanks.
